### ISSUE TITLE
`isPlainObject` (and others?) impacted by breaking changes in typescript 5.4

### ISSUE THREAD

**Issue by stwlam** [remeda/remeda#559]
I'm noticing some unexpected breaks in an existing codebase that uses remeda 1.47.0 when upgrading to typescript 5.4. Small reproduction case [playground link](https://www.typescriptlang.org/play?ts=5.4.1-rc#code/C4TwDgpgBAcghgJwQewO4QCYBVkB4sA0UAQnAM4QB8UAvFAKIAewCcAxsPkaRdRMxAB2GMlEEQAbhAQAoKFAD8JchDlQAXA2asOXZbwDcMmRghsANomgAzAK6COAS2SCojsgAVLjwQHkARgBWZpxYlAAUahhwwHCaWFAAPlAASmbICBi4HiiQCKAA0hAgRPYA1oJogpQEMgCUmtGxbqLwSGiYOHppbBlZOch5hcWlghVVlJRGJmaWCNC9gmTAUAC2IAGBmuWVqIJGjtZQ4e5ecD6bIeHrm3V1UADeavLWGceLy1BlxVDIRwDaACI4ICiID-KCoIC2ICALr3J7yJFQD4rUwQPK0NYbIL-b4gWFGZHyVFQdF5IogUR0U7ePxBK7k6T3JSXDgAOnxZHCTIQ900-0JzygAF8ZGKgA):

```ts
type NarrowedTo<T, Base> = Extract<T, Base> extends never
  ? Base
  : Extract<T, Base>;

declare function isPlainObject<T>(
  data: T | Record<PropertyKey, unknown>,
): data is NarrowedTo<T, Record<PropertyKey, unknown>>;

declare const myObj: unknown;
if (isPlainObject(myObj)) {
    for (const key of ["a", "b", "c"]) {
      const deeper = myObj[key];
      const deeperKeys = isPlainObject(deeper) ? Object.keys(deeper) : [];
    }
}
```

The playground link uses typescript 5.4.1-rc, but the issue is also present in 5.4.2.

**Comment by TkDodo** [comment]
Interesting, thanks for the reproduction. If you know how to fix it, would you contribute a PR ?

**Comment by cjquines** [comment]
i think there are some language changes in 5.4; when i tried upgrading to ts 5.4 there were a handful of errors relating to objects

**Comment by stwlam** [comment]
> Interesting, thanks for the reproduction. If you know how to fix it, would you contribute a PR ?

I'm not sure what the source of this issue is but will investigate after the week is over. For what it's worth, it's narrowable to a change made after 5.4 beta: [playground link](https://www.staging-typescript.org/play?ts=5.4.0-beta#code/C4TwDgpgBAcghgJwQewO4QCYBVkB4sA0UAQnAM4QB8UAvFAKIAewCcAxsPkaRdRMxAB2GMlEEQAbhAQAoKFAD8JchDlQAXA2asOXZbwDcMmRghsANomgAzAK6COAS2SCojsgAVLjwQHkARgBWZpxYlAAUahhwwHCaWFAAPlAASmbICBi4HiiQCKAA0hAgRPYA1oJogpQEMgCUmtGxbqLwSGiYOHppbBlZOch5hcWlghVVlJRGJmaWCNC9gmTAUAC2IAGBmuWVqIJGjtZQ4e5ecD6bIeHrm3V1UADeavLWGceLy1BlxVDIRwDaACI4ICiID-KCoIC2ICALr3J7yJFQD4rUwQPK0NYbIL-b4gWFGZHyVFQdF5IogUR0U7ePxBK7k6T3JSXDgAOnxZHCTIQ900-0JzygAF8ZGKgA)

**Comment by eranhirsch** [comment]
Seems like a regression in typescript, even removing most of the complexity it still fails with a generic and succeeds without one:

https://www.typescriptlang.org/play?#code/PTAEF5NB1B5AlA0gZQpdH0CgsBMCmAxgDYCGATvqAGYCuAdoQC4CWA9vaCwM4AKZLerABGAKyJMAjAAosoULlJNSALlAMA1vTYB3egBosASjWLlXbqHhE25XAB5e5NgAd85JgE9E+T-vX0Wrr0AHwA3DgEJBRUhBzcTKAAtp4iopJqmtp6ESzUoNI8-KSCaRIyKWmSRkagAN5y8jS2BXH0CaAavqBs+QDaAESkA-4DwiOgA4QDALq1DU1NbR0E+G7kEMmpYpJ9XZ4zEYvyIIsAegD8jYvLiavrPp6W4BbFpWLl0vfutRegZcwAHT7bhffBrH6gNR9Q7XUAAXywiKwp3QoAAIrAAKLIAByABUYAhEGhMGTsKdYPRiJ4FHlqO58IwqMJ8EwdODOEwABZUABMXCSLmIlh4oB5VAA5kz3CxCKBCYJxbzxZ43P58AA3JmgHS8zgsRI8egAckStG4+FwgJtkSIZEoNAYzHYBr4AiEH2YfPs+JCsnkZlUASCekMJgUSlIFisNjsjmc6y8j38WWC4Tt0Udty2aT5mUC2XouXyhXdJU94m90kqYj5NXqcOoLWkOf2PX6QwmY270zmjeOoBz3w2L1roj5e18sMHpyalzh8mH4Ievmerw9AKYfLBEPIv3+XqYwLXu-WtWhM6aiMRQA

**Comment by eranhirsch** [comment]
The TypeScript team labeled this issue as their bug, meaning we cannot fix it on our side. If you run into issues, I suggest not updating TypeScript to 5.4 for now. Follow along in the attached task.

**Comment by eranhirsch** [comment]
The typescript issue was fixed with a revert merged into version 5.4, but I don't know how the release cycle works. Once we can confirm this is fixed in a later version of 5.4 (I think it still happens on version 5.4.2), I'll close this issue when that version comes out.
